src/DynamoPackages/PackageLoader.cs

```diff
@@ -678,36 +678,54 @@ private void CheckPackageNodeLibraryCertificates(string packageDirectoryPath, Pa
-                    assem = mlc.LoadFromAssemblyPath(filename);
-                    var mlcAssemblies2 = mlc.GetAssemblies();
-                    //if loading the assembly did not actually add a new assembly to the MLC
-                    //then we've loaded it already, and our current behavior is to
-                    //disable publish when a package contains the same assembly twice.
-                    if (mlcAssemblies2.Count() == mlcAssemblies.Count())
-                    {
-                        throw new FileLoadException(filename);
-                    }
-                    return AssemblyLoadingState.Success;
-                }
-                catch (BadImageFormatException)
-                {
-                    assem = null;
-                    return AssemblyLoadingState.NotManagedAssembly;
-                }
-                catch (FileLoadException)
+            Assembly assemName = null;
+            assem = null;
+            try
+            {
+                var mlcAssemblies = mlc.GetAssemblies();
+                assemName = mlcAssemblies.FirstOrDefault(x => x.GetName().Name.ToLower().Equals(Path.GetFileNameWithoutExtension(filename).ToLower()), null);
+                assem = mlc.LoadFromAssemblyPath(filename);
+                
+                var mlcAssemblies2 = mlc.GetAssemblies();
+                //if loading the assembly did not actually add a new assembly to the MLC
+                //then we've loaded it already, and our current behavior is to
+                //disable publish when a package contains the same assembly twice.
+                if (mlcAssemblies2.Count() == mlcAssemblies.Count())
                 {
-                    assem = null;
-                    return AssemblyLoadingState.AlreadyLoaded;
+                    throw new FileLoadException(filename);
                 }
+                return AssemblyLoadingState.Success;
+            }
+            catch (BadImageFormatException)
+            {
+                assem = null;
+                return AssemblyLoadingState.NotManagedAssembly;
+            }
+            catch (FileLoadException)
+            {
+                assem = assem == null ? assemName : assem;
+                return AssemblyLoadingState.AlreadyLoaded;
+            }
+        }
+
+        internal static MetadataLoadContext InitSharedPublishLoadContext()
+        {
+            // Retrieve the location of the assembly and the referenced assemblies used by the domain
+            var runtimeAssemblies = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
+            // Create PathAssemblyResolver that can resolve assemblies using the created list.
+            var resolver = new PathAssemblyResolver(runtimeAssemblies);
+            return new MetadataLoadContext(resolver);
+        }
+        internal static void CleanSharedPublishLoadContext(MetadataLoadContext mlc)
+        {
+            mlc.Dispose();
         }
```

Comment:
@zeusongit, so as discussed last week, I'm not sure if disposing the MLC is even necessary (at least in a publish new version scenario) as the assemblies that need to be deleted would be loaded into Dynamo at the time of publishing and would therefore not be accessible for deletion irrespective of the fact they are loaded into the MLC or not.

Reply:
The MLC still need to be disposed to fix another bug where if a user re-tries to publish another version in the same session they will encounter jumbled tree view due to MLC still pointing to the older assemblies.

Reply:
Ok, I still feel we should clean this up in the future where we only load new assemblies (for the new package we wish to upload) into the MLC for inspection and continue to load existing assemblies if they are unchanged.